### PR TITLE
Add plotting n/n

### LINKED ISSUE #1406: Add `.plot()` to all modular metrics
## 🚀 Feature

We are in the process of adding native plotting to torchmetrics such that our users can easily create simple plots of metrics, which is especially useful for metrics such as confusion matrix and pr-curves but even single-scalar metrics such as accuracy it can be beneficial to quickly create a plot of computed values.

The first proof of concept PR has just been merged: https://github.com/Lightning-AI/metrics/pull/1328

We are inviting anyone interested in contributing to torchmetrics to help add this feature to all remaining metrics. The tasks are fairly simple and an example of how to implement the `plot` metric can be seen here: https://github.com/Lightning-AI/metrics/blob/d8caeeb28efc0afb86ae68cf66b06412749a4e46/src/torchmetrics/classification/accuracy.py#L110-L156

We would like the work divided into PRs tackling each domain, however if any are interested in only contributing to a few metrics within a particular domain that is also more than welcome.

- [ ] Aggregation ->#1485 @SkafteNicki 
- [ ] Audio ->#1434 @shhs29
- [ ] Classification -> @srishti-git1110
- [ ] Image ->#1480 @venomouscyanide
- [ ] Detection
- [ ] Regression -> @srishti-git1110
- [ ] Retrieval -> @ufukhurriyetoglu
- [ ] Text -> @ufukhurriyetoglu

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #1406
Final PR for plotting (for now). Adds:
* documentation page + examples
* `plot` method to `MetricCollection` + test
* smaller fixes to other metrics

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/pages/plotting.rst [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+    import matplotlib
+    import matplotlib.pyplot as plt
+    import torchmetrics
+
+########
+Plotting
+########
+
+.. note::
+    The visualzation/plotting interface of Torchmetrics requires ``matplotlib`` to be installed. Install with either
```
Comment (stancld) [inline docs/source/pages/plotting.rst:13]:
```suggestion
    The visualization/plotting interface of Torchmetrics requires ``matplotlib`` to be installed. Install with either
```

File: docs/source/pages/plotting.rst [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+    ``pip install matplotlib`` or ``pip install 'torchmetrics[visual]'``. If the ladder option is chosen the
```
Comment (stancld) [inline docs/source/pages/plotting.rst:14]:
```suggestion
    ``pip install matplotlib`` or ``pip install 'torchmetrics[visual]'``. If the latter option is chosen the
```

File: docs/source/pages/plotting.rst [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+.. code-block:: python
+
+    metric = AnyMetricYouLike()
+    for _ in range(num_updates):
+        metric.update(preds[i], target[i])
+    fig, ax = metric.plot()
+
+``.plot`` will always return two objects: ``fig`` is an instance of :class:`~matplotlib.figure.Figure` which contains
+figure level attributes and `ax` is an instance of :class:`~matplotlib.axes.Axes` that contains all the elements of the
+plot. These two objects allow to change attributes of the plot after it is created. For example if you wanted to make
```
Comment (stancld) [inline docs/source/pages/plotting.rst:30]:
```suggestion
plot. These two objects allow to change attributes of the plot after it is created. For example, if you want to make
```

File: docs/source/pages/plotting.rst [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+    collection.plot(val=values, ax=ax)
+
+.. image:: ../_static/images/binary_accuracy_multistep.png
+   :height: 100px
+   :width: 200 px
+   :scale: 50 %
+   :alt: multistep accuracy plot
+   :align: right
+
+However, the ``plot`` method of ``MetricCollection`` also support an additional argument called ``together`` that will
```
Comment (stancld) [inline docs/source/pages/plotting.rst:194]:
```suggestion
However, the ``plot`` method of ``MetricCollection`` also supports an additional argument called ``together`` that will
```

**Review by stancld — APPROVED** [review]
This is golden, thanks a lot! <3

File: docs/source/pages/plotting.rst [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+   :align: right
+
+Here, each element is assumed to be an independent metric and plotted as its own point for comparing. The above is true
+for all metrics that returns a scalar tensor, but if the metric returns a tensor with multiple elements then the
+``.plot`` method will return a specialized plot for that particular metric. Take for example the ``ConfusionMatrix``
+metric:
+
+.. code-block:: python
+
+    metric = torchmetrics.ConfusionMatrix(task="multiclass", num_classes=3)
```
Comment (Borda) [inline docs/source/pages/plotting.rst:107]: thinking about using inline plotting?
So a user could also just copy the code and get the same results...

**Review by Borda — APPROVED** [review]

File: docs/source/_static/images/tracker_binary.png
Comment (justusschock) [inline docs/source/_static/images/tracker_binary.png:1]: is there a way to dynamically create those images for docs? Ideally I'd not want them to live in the repo to avoid history from growing (this is true for all binary files). Otherwise we can probably move them to s3 cc @Borda

**Review by justusschock — APPROVED** [review]

File: docs/source/_static/images/tracker_binary.png
Comment (Borda) [reply docs/source/_static/images/tracker_binary.png:1]: already asked in https://github.com/Lightning-AI/torchmetrics/pull/1682#discussion_r1156141850 :chipmunk:

**Comment by SkafteNicki** [comment]
@Borda, @justusschock plots should be dynamically created now :)

File: docs/source/pages/plotting.rst [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+could be do like this:
+
+.. code-block:: python
+
+    ax.set_fontsize(fs=20)
+    fig.set_title("This is a nice plot")
+    fig.save_fig("my_awesome_plot.png")
+
+If you want to include a Torchmetrics plot in a bigger figure that have subfigures and subaxises all ``.plot`` methods
+support an optional `ax` argument where you can pass in the subaxises you want the plot to be inserted into:
```
Comment (awaelchli) [inline docs/source/pages/plotting.rst:41]:
```suggestion
If you want to include a Torchmetrics plot in a bigger figure that has subfigures and subaxes, all ``.plot`` methods
support an optional `ax` argument where you can pass in the subaxes you want the plot to be inserted into:
```

File: docs/source/pages/plotting.rst [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+        metric.update(torch.rand(10,), torch.randint(2, (10,)))
+    fig, ax = metric.plot()
+
+.. plot:: pyplots/binary_accuracy.py
+    :scale: 100
+    :include-source: false
+
+A single point plot is not that informative in itself, but if available we will try to include additional information
+such as the lower and upper bounds the particular metric can take an if the metric should be minimized or maximized
+to be optimal. This is true for all metrics that return a scalar tensor.
```
Comment (awaelchli) [inline docs/source/pages/plotting.rst:79]:
```suggestion
such as the lower and upper bounds the particular metric can take and if the metric should be minimized or maximized
to be optimal. This is true for all metrics that return a scalar tensor.
```

File: docs/source/pages/plotting.rst [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+Some metrics returns multiple values (such as an tensor with multiple elements or an dict of scalar tensors), and in
+that case calling ``.plot`` will return a figure similar to this:
```
Comment (awaelchli) [inline docs/source/pages/plotting.rst:81]:
```suggestion
Some metrics return multiple values (such as an tensor with multiple elements or a dict of scalar tensors), and in
that case calling ``.plot`` will return a figure similar to this:
```

File: docs/source/pages/plotting.rst [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+    for _ in range(num_updates):
+        metric.update(torch.randint(3, (10,)), torch.randint(3, (10,)))
+    fig, ax = metric.plot()
+
+.. plot:: pyplots/multiclass_accuracy.py
+    :scale: 100
+    :include-source: false
+
+Here, each element is assumed to be an independent metric and plotted as its own point for comparing. The above is true
+for all metrics that returns a scalar tensor, but if the metric returns a tensor with multiple elements then the
```
Comment (awaelchli) [inline docs/source/pages/plotting.rst:95]:
```suggestion
for all metrics that return a scalar tensor, but if the metric returns a tensor with multiple elements then the
```

File: docs/source/pages/plotting.rst [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+    )
+    fig, ax = plot_class.plot(value)
+
+********************
+Plotting multi steps
+********************
+
+In the above examples we have only plotted a single step/single value, but it is also possible to plot multiple steps
+from the same metric. This is often the case when training a machine learning model, where you are tracking one or
+metrics that you want to plot as they are changing over time. This can be done by provided a sequence of outputs from
```
Comment (awaelchli) [inline docs/source/pages/plotting.rst:127]:
```suggestion
from the same metric. This is often the case when training a machine learning model, where you are tracking one or
multiple metrics that you want to plot as they are changing over time. This can be done by providing a sequence of outputs from
```
was the "multiple" missing?

File: docs/source/pages/plotting.rst [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+        collection.reset()
+    collection.plot(val=values, ax=ax)
+
+.. plot:: pyplots/binary_accuracy_multistep.py
+    :scale: 100
+    :include-source: false
+
+However, the ``plot`` method of ``MetricCollection`` also supports an additional argument called ``together`` that will
+automatically try to plot all the metrics in the collection together in the same plot (with appropriate labels). This
+is only possible if all the metrics in the collection returns a scalar tensor.
```
Comment (awaelchli) [inline docs/source/pages/plotting.rst:181]:
```suggestion
is only possible if all the metrics in the collection return a scalar tensor.
```

**Review by awaelchli — APPROVED** [review]
I reviewed the docs part and pointed out minor typos. 
Very nice guide!
